File: apps/framework-internal-app/app/flows/UserActivity/ParsedActivity/flow.ts
// Example flow function: Converts local timestamps in UserActivity data to UTC.

// Imports: Source (UserActivity) and Destination (ParsedActivity) data models.
import { ParsedActivity, UserActivity } from "../../../datamodels/models";

// The 'run' function transforms UserActivity data to ParsedActivity format.
// For more details on how Moose flows work, see: https://docs.moosejs.com
export default function run(event: UserActivity): ParsedActivity {
  // Convert local timestamp to UTC and return new ParsedActivity object.
  return {
    eventId: event.eventId, // Retain original event ID.
    userId: "puid2" + event.userId, // Example: Prefix user ID.
    activity: event.activity, // Copy activity unchanged.
    timestamp: new Date(event.timestamp.toUTCString()), // Convert timestamp to UTC.
  };
}

File: apps/framework-cli/tests/python/flows/valid/flow.py
from dataclasses import dataclass
from typing import Callable
from datetime import datetime

@dataclass
class Flow:
    run: Callable

type Key[T: (str, int)] = T

@dataclass
class UserActivity:
    eventId: Key[str]
    timestamp: str
    userId: str
    activity: str

@dataclass
class ParsedActivity:
    eventId: Key[str]
    timestamp: datetime
    userId: str
    activity: str

def my_func(dm: UserActivity) -> ParsedActivity:
    return ParsedActivity(
        eventId=dm.eventId,
        timestamp=datetime.fromisoformat(dm.timestamp),
        userId=dm.userId,
        activity="yo"
    )

my_flow = Flow(
    run=my_func
)

